### ISSUE TITLE
[API Proposal]: Create array from array type

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#76478]
### Background and motivation

The existing reflection APIs for creating array instances take array element type. If one has the actual array type available, this is both inefficient (requires going from the array type to the element type and back to the array type that we have started with) and requires suppressing warnings about the array type potentially not being available with Aot.

The proposal is the introduce a set of APIs that mirror the existing APIs for creating array instances, except that they take the array type instead of the element type.

### API Proposal

```csharp
namespace System;

public class Array
{
        // Proposed APIs
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2, int length3);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params int[] lengths);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int[] lengths, int[] lowerBounds);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params long[] lengths);

        // Existing APIs
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("The code for an array of the specified type might not be available.")]
        public static System.Array CreateInstance(System.Type elementType, int length);
        public static System.Array CreateInstance(System.Type elementType, int length1, int length2);
        public static System.Array CreateInstance(System.Type elementType, int length1, int length2, int length3);
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("The code for an array of the specified type might not be available.")]
        public static System.Array CreateInstance(System.Type elementType, params int[] lengths);
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("The code for an array of the specified type might not be available.")]
        public static System.Array CreateInstance(System.Type elementType, int[] lengths, int[] lowerBounds);
        [System.Diagnostics.CodeAnalysis.RequiresDynamicCode("The code for an array of the specified type might not be available.")]
        public static System.Array CreateInstance(System.Type elementType, params long[] lengths);
}
```

### API Usage

We can replace `Array.CreateInstance(arrayType.GetElementType()!, length)` pattern in number of places in dotnet/runtime libraries with `Array.CreateInstanceFromArrayType(arrayType, length)`:

- https://github.com/dotnet/runtime/blob/e461b776bff0e5a6b8c72282a9f66695a837461d/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/SurrogateDataContract.cs#L71
- https://github.com/dotnet/runtime/blob/983c8f239a98812498d874ed36b7001bd764fdfe/src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs#L2031
- https://github.com/dotnet/runtime/pull/76473#discussion_r985033231

### Alternative Designs

Introduce overload for single dimensional arrays only. The single dimensional arrays are likely the 99+% use case for this API.

**Comment by jkotas** [comment]
cc @dotnet/ilc-contrib Does this look reasonable?

**Comment by teo-tsirpanis** [comment]
@jkotas can you set a milestone? "Untriaged" means without one. I assume this is not for 7.0.0.

**Comment by MichalStrehovsky** [comment]
> cc https://github.com/orgs/dotnet/teams/ilc-contrib Does this look reasonable?

Looks reasonable to me! I'll leave it up to the API review board to think about the naming or the location of the API (we could also consider stashing it somewhere else, such as on Activator (add another overload of CreateInstance?), or RuntimeHelpers (GetUninitializedArray?).

**Comment by bartonjs** [comment]
```C#
namespace System;

public partial class Array
{
        // Proposed APIs
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2, int length3);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params int[] lengths);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int[] lengths, int[] lowerBounds);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params long[] lengths);
}
```

**Comment by AlexRadch** [comment]
> The implementer is encouraged to consider whether all of these overloads are actually required, as the suggestion that the first overload will be in excess of 99% of all calls seems correct.

@bartonjs I decided to implement only the next overloads:

```cs
namespace System;

public partial class Array
{
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params int[] lengths);
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int[] lengths, int[] lowerBounds);
}
```

And I have not implemented the following overloads:

```cs
namespace System;

public partial class Array
{
        // It is not used in the current code and overload with `params int[] lengths` can be used instead
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2); 

        // It is not used in the current code and overload with `params int[] lengths` can be used instead
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, int length1, int length2, int length3);

        // Array does not support lengths more than `int.Max`. Array always uses `int` for lengths, not `long`
        public static System.Array CreateInstanceFromArrayType(System.Type arrayType, params long[] lengths);
}
```

**Comment by tannergooding** [comment]
Not going to land in .NET 8. This can effectively be done at any time after main opens for .NET 9 next month

**Comment by jkotas** [comment]
Fixed by #88620
